src/processing/UndoRedo.java

```diff
@@ -0,0 +1,245 @@
+package processing;
+
+import processing.*;
+import processing.utility.*;
+import processing.boardobject.*;
+import java.util.*;
+
+/**
+ * UndoRedo class implements undo-redo operations
+ * 
+ * @author Rakesh Kumar
+ * @reviewer Devansh Singh Rathore
+ */
+
+public class UndoRedo {
+	
+	/** maximum stack size 
+	 *  restricts the maximum undo-redo operations
+	 */
+	private static int STACK_CAPACITY = 20;
+	private enum Operation {
+		UNDO, REDO;
+	}
+    
+	/**
+	 *  Will be visited later
+	 *  map for access in the list
+   	 *  private static Map <ObjectId, ListIterator> idToIterator;
+     */
+	
+	/**
+	 * This helper method will push the object into the given stack
+	 * 
+	 * @param stack could be undo or redo stack
+	 * @param obj object to be pushed
+	 */
+	private static void addIntoStack(LinkedList <BoardObject> stack, BoardObject obj) {
+		
+		/** If stack size becomes full, delete the bottom (first) object */
+		if (stack.size() >= STACK_CAPACITY)
+			stack.removeFirst();
+		
+		/** Push the object to the top (last) */
+		stack.addLast(obj);
+	}
+	
+	/**
+	 * This helper method calls drawCure method of CurveBuilder class to create the object
+	 * It is called during undo-redo operations
+	 * 
+	 * @param obj
+	 * @return newly created object
+	 */
+	private static BoardObject createOperation(BoardObject obj) {
+		/** BoardObject CREATE operation*/
+		IBoardObjectOperation newOp = new CreateOperation();
+		/** gets the newly created object */
+		BoardObject newObj = CurveBuilder.drawCurve( 
+									obj.getPixels(), 
+									newOp, 
+									obj.getObjectId(), 
+									obj.getTimestamp(),
+									obj.getUserId(), 
+									obj.getPrevIntensity(), 
+									false 
+									);
```

Comment:
if there already exists a `Boardobject obj`, then what is the need for calling `CurveBuilder.drawCurve()` to create a new one? Please clarify, I might be wrong in this case.

Reply:
Because otherwise you can just directly set the `newOp` using `BoardObject` functions and that would save a lot of time complexity.

Reply:
CurveBuilder creates an object by assigning references. There is no actual copying of the data members so, time complexity will be not high. 
I need to have two copies (operation copies). If I set the newOp then it's previous operation would be lost and I can't further undo or redo it.